### ISSUE TITLE
[GSOC] User Interface improvements proposal-mohamedAdhamc

### ISSUE THREAD

**Issue by mohamedAdhamc** [synfig/synfig#2611]
As mentioned in the docs I'm starting this issue to discuss my application and get some feed-back. I'm not sure if this is still the preferred method of doing so, if not kindly inform me so and my apologies in advance. @ice0 , @morevnaproject  you're the mentors for this project I am not sure exactly who of you will be or maybe both ? 😅. I do have a few questions 

1- where should I submit the proposal for review? should it be through the official website for GSoC or is there another way preferred for doing so?
2- do I have to wait till the proposal is completely done to send it in for review or could I send it  while in progress over the coming days.?
That's it for now, thanks in advance 😃.

**Comment by ice0** [comment]
Here we discussing technical things mostly, just in case to not overload contributors who subscribed to all messages here. 
Answered on the forums.

**Comment by mohamedAdhamc** [comment]
> Here we discussing technical things mostly, just in case to not overload contributors who subscribed to all messages here. Answered on the forums.
 
Alright, sounds good.

**Comment by mohamedAdhamc** [comment]
Hey guys, @ice0 @rodolforg ,and anyone else who might sees this (^_^) 

So I was looking into the first feature which is #2528 , "Add the posibility of move the paramether values using the mouse"
now this was the link provided that described how it could possibly look like https://imgur.com/a/fRek4Rm .

I see a few problems here and I hope you guys could maybe give me some insight on them
1- For some parameters like the ones of type "real", and "int". the current implementation uses spin buttons, and they already have the option to increase or decrease using the scroll wheel and the up/down button, so would this option be possibly redundant for them? or is it okay

2- There would be a conflict of selection/highlighting the value so that it is copied/pasted etc... and between the ability to click and then swipe the mouse right or left to increase or decrease ( that is the way to select the value ).

hope to get some insight from you guys on what you think.

**Comment by ice0** [comment]
> 1- For some parameters like the ones of type "real", and "int". the current implementation uses spin buttons, and they already have the option to increase or decrease using the scroll wheel and the up/down key, so would this option be possibly redundant for them? or is it okay

I think it's better to have the same UX for all widgets.

> 2- There would be a conflict of selection/highlighting the value so that it is copied/pasted etc... and between the ability to click and then swipe the mouse right or left to increase or decrease ( that is the way to select the value ).

There is no conflict. For example in Adobe products, to change the value using mouse you should use mouse-down (not click) and move. If you click it will go into edit mode so you can copy/paste/type the value manually. In Synfig, mouse down event forces edit mode, but you can temporarily ignore this and check mouse down event on the "Name" column.

> 3- Also would overriding the spinbutton here make sense I believe the spin button is kind of standard so overriding it (other than the technical problems) might be confusing/weird for the users themselves which I would assume a lot of them would be using other gtk apps as well.

You can ignore this for now, we will back to this later, after behavior will be implemented on the "Name" column.

**Comment by ice0** [comment]
> Oh ok so first I should do this for the name column instead correct?

Yes. In this case, we are extending the current UX without changing it.

**Comment by mohamedAdhamc** [comment]
#1936 is feature number 2. I just have a few question before doing it @ice0 
1- This is also for animation tabs only correct ?
2- I think it also makes sense to implement this in the same way as the other feature correct ?

**Comment by ice0** [comment]
> #1936 is feature number 2. I just have a few question before doing it @ice0 
> 1- This is also for animation tabs only correct ? 

Better to try it to implement for all docks windows.

> 2- I think it also makes sense to implement this in the same way as the other feature correct ?

Yes.

**Comment by mohamedAdhamc** [comment]
@ice0 
as per requested here is the daily summary today:

time: started at around 5 pm Cairo time and now ended at 2 AM cairo time. However it wasnt continous meaning maybe two hours I was eating etc.. Also I think maybe bare minimum 90 mins were lost in build process. whenever I make a slight change in any header it takes about 5-25 mins to rebuild depending on the file.

description: 
1- I was working on feature 2 (switching tabs using ctr+tab) : didn't stay too long doing this as I went on to point 2 in the middle. but a quick description of what happened there so far. Well, I had implemented feature 2 by overiding the on_key_pressed call back of notebook class inside dockbook class. But now since the update was that I had to implement it as an action so that the shortcut can be edited using the shortcut dialog I was facing a problem in which that I had to synchronize the action with the dock books and have it only switch the page of the selected dock book which could be done in many ways
the way I was thinking of in the morning was to have the action set a flag and then the on_key_pressed doing the work as usual. Anyway, I couldn't really continue much as I move on to point 2.

2- I was adressing the comments on feature number 3 I was kind of lost by what you meant with the arguments until I realized you meant that the issue was with what i now know is called the get_text(). I spent a bit of time trying to understand which part exactly and for the back and forth commnication. then I read the links you sent which were very helpful btw, and made a commit with the new changes.

3- this was probably the biggest chunk of the day. working on trying to fix the history pollution resulting from feature one. I spent a long time going around the codebase trying to find the part responsible for dealing with setting the layer parameter action. I went through multiple files/classes like  layerparmset, historytreestore, dock_history, and other which I do not remember. then after that i found a specific function which is responsible for adding the actions to the history treeview and then a very simple idea came into my mind, use a flag which is set when the parameter is being edited by the mouse.
-when the parameter is being edited by the mouse don't show the action,
-when it is not being edited by mouse (e.g. bykeyboard,spinbutton, etc...)show actions normally. 
-when the mouse editing is just done then show the very last action.
along the way as I was trying multiple things in different files build time was slowing me down a bit --- _which is why I was wondering if I move over to Linux maybe then I could save the time but then I also have ubuntu 32 bit verision installed so I need to reinstall but  64 bit to be able to use certain tools so its a bit of a hassle to move and I brueshed the idea aside to when I have more free time maybe_---
then I spent a little time cleaning the PR(removing comments, unused functions,etc...) and testing and finally I pushed the commit and now writing this.


Yesterday: 
while I don't remember with many details like today what I was working on. I was working on feature 2 I believe mostly and it was very similar to point one today. at the end of the day I think I had it switching tabs for the selected dockbook but there were logical errors and other errors which needed to be adressed.

p.s. comparing the time of the last commit and now, writing this took about 35 minutes to try to remember and describe  in detail what i was doin exactly. I don't know if it would be necessary to do these everyday but maybe if it's good I could just try to put main bullet points of what I did instead of writing in detail like this to save time.

**Comment by mohamedAdhamc** [comment]
ok quick general project update 
- now major functionality of feature one is done 

During the next two days, I will be trying to get feature 2 "ctr+tab to switch tabs" done.
Also along with responding to the reviews on feature one and the possible enhancements/optimizations.

starting on the 20th if things go to plan I will be seeing what's left over in feature 4(show/hide ruler) to do ( this one was actually done before the coding period which I'm now glad for xD) but it was left with a question if the button is needed or just the menu button was enough.

**Comment by ice0** [comment]
> During the next two days, I will be trying to get feature 2 "ctr+tab to switch tabs" done. Also along with responding to the reviews on feature one and the possible enhancements/optimizations.

Can you explain what you think will take the most time here?

**Comment by mohamedAdhamc** [comment]
> 
> Can you explain what you think will take the most time here?

Well I'm not quite sure how it should be implemented now that it should be done as an action. I thought maybe make the action set a flag and if this flag is set have the override of the on_key_pressed work. But this can't really work as then what ends up happening is that after the action is done pressing any key will switch tabs. Also if I just make an action for each dockbook I would have to make multiple actions so that one is no good. I think the best way now would be to just do the action on whichever dockbook is in focus. How to get the dockbook in focus, is the question. 

P.s. CTR+ tab isn't accepted in shortcut dialog as an accelerator, it give out a warning on running the program. And on my windows machine I couldn't even manually set it to CTR+tab.

**Comment by mohamedAdhamc** [comment]
Quick General update:
feature 1 : main things left are history pollution and escape to cancel
feature 2: waiting on review 
feature 3: I think this one should be done there were a couple things i resolved from reviews. so I guess we could also say 
                waiting on review here too to see if its good to go.
feature 4:  I will remove the button for this one and ust leave the menu item and it would be also then same status as 3.

what I will be doing next 2 days:
1- remove the button in feature 4
2- apply the same method of fixing history pollution in color wheel to feature 1
3- allow for esc to cancel change in feature 1
4- address the possible things requested in any of the four features.

General schedule review:
According to the schedule, the first four have until the 30th. But I think it would be wiser to advance on to the next features as quickly as possible.

**Comment by mohamedAdhamc** [comment]
Hey guys, @ice0 , @rodolforg  hope you're doing well !

I kindly suggest if possible you guys review #2731 and #2634  first as these will probably take the least time to do so and they have already been reviewed and resolved a couple of time so we can get them out of the way.#2755 should also not take too much time. then finally we have #2717  this one will probably take the longest to test.

Also, do you guys think it's a good idea to start on the fifth feature along the way? I mean there may be things to resolve but after I resolve them there will be time left in the day, then I'll use it for the fifth feature.

**Comment by mohamedAdhamc** [comment]
ok, quick general progress update:

as of now, following features are done but still under different stages of the review phase as follows.
feature 1: #2717 , last thing ice0 said it will take some time for him to test a certain part in a test app.
feature 2: #2731 , this one is fully reviewed
feature 3: #2755 , not yet reviewed but shouldn't take long to review.
feature 4: #2634 , ice0 said he also needs to test something here.
feature 5: #2763 , not yet reviewed.

schedule position: according to the schedule feature 5 had until august 6. so I think as of now things are going quite well.

what I will be doing for the rest of the project time:
 This project is a bit on the longer and more discrete side. It includes 12 features which vary in complexity and breakablity. so what ill be doing is to work on the next feature as soon as I finish the one before it, and whenever something is mentioned from the reviews will go back, address it then resume.

what I will be doing next 2/3 days:
-get things started with feature 6
-address things from reviews.

**Comment by mohamedAdhamc** [comment]
Hey everyone :)

progress update:
as of now, the following features are done but still under different stages of the review phase as follows.
feature 1: https://github.com/synfig/synfig/pull/2717 , last thing ice0 said it will take some time for him to test a certain part in a test app.
feature 2: https://github.com/synfig/synfig/pull/2731 , this one is fully reviewed
feature 3: https://github.com/synfig/synfig/pull/2755 ,  - ice0 suggested waiting till a decision is made on whether we should 
------------------------------------------------------------- get around gtk not allowing tab to be used as an accelerator or not.
feature 4: https://github.com/synfig/synfig/pull/2634 ,  - ice0 said he also needs to test something here.
feature 5: https://github.com/synfig/synfig/pull/2763 , - not yet reviewed.
feature 6: https://github.com/synfig/synfig/pull/2765 - review in progress
feature 7: #2767 not yet reviewed

schedule position:
- according to schedule feature 7 had until august 20th.
- these seven features were the "main" features of the project
- now remains 5 "additional" features (one of which was actually almost finished before coding period)
now just throwing out an approximate date I think maybe the remaining features could be done by august 16th maybe. then we'd have from August 16th till the end October 3rd to enhance/address things related to the features (which is also done now).

**Comment by mohamedAdhamc** [comment]
Alright really quick update here, I feel like the effectiveness has gone really down maybe because I have taken a complete day off in very long. So today I'll try not to do anything at all. Then hopefully tomorrow I'll be back to pace.

Feature 8 just needs a bit more cleaning up and it will be ready for review then I'll be advancing onto the remaining features (4, one of which I did before the coding period).

**Comment by mohamedAdhamc** [comment]
Hey everyone, long time no update xD. 

"rotating guides" feature was a little bigger than expected, and there were many review so I had to kind of wait a bit for this update :).

progress update:
as of now, the following features are done but still under different stages of the review phase as follows.
feature 1: https://github.com/synfig/synfig/pull/2717 : seems like the biggest thing here as of now is the history pollution which is being discussed in another pr #2750.
feature 2: https://github.com/synfig/synfig/pull/2731 : fully reviewed
feature 3: https://github.com/synfig/synfig/pull/2755 : ice0 suggested waiting till a decision is made on whether we should
------------------------------------------------------------- get around gtk not allowing tab to be used as an accelerator or not.
feature 4: https://github.com/synfig/synfig/pull/2634  : reviewed a couple times by ice0 and rodolforg. "almost there" as stated -------------------------------------------------------------- by rodolf :)
feature 5: https://github.com/synfig/synfig/pull/2763  : reviewed once, and was "initial review" by rodolforg
feature 6: https://github.com/synfig/synfig/pull/2765 : This one is close to done, but there are certain parts that need                
 ------------------------------------------------------------ opinions (around 2 points). as stated by rodolf
------------------------------------------------------------ " One thing that bothers me is that the current time changes when          -------------------------------------------------------------dragging the playback boundaries.
-------------------------------------------------------------But I don't know if it is really a problem or just me XD
-------------------------------------------------------------The rest seems to work nice :) "
feature 7: https://github.com/synfig/synfig/pull/2767 : I believe review by rodolforg is done and as said by him " I believe the    -------------------------------------------------------------only points left (or not) are those I'm waiting for ice0 to                        
-------------------------------------------------------------comment/review.
-------------------------------------------------------------I believe the only points left (or not) are those I'm waiting for ice0 to ----------------------------------------------------------------comment/review. "
feature 8: https://github.com/synfig/synfig/pull/2777: ready for review.
feature 9: https://github.com/synfig/synfig/pull/2805: ready for review.
feature 11: https://github.com/synfig/synfig/pull/2615: reviewed couple times last thing ice0 said to rebase and do a certain ----------------------------------------------------------------thing so I'm assuming its very close to being fully reviewed (if not), but --------------------------------------------------------------there might be a slight issue with the separator

schedule position:
-according to the schedule ten features have until September 12th.
- what remains now are 2 features which I will hopefully try to get done as quick as possible so that all time left till october 3rd is left for reviews and "self enhancing" the features. :)

**Comment by mohamedAdhamc** [comment]
Actually I should have written this couple days back xD but I was caught up in the history pollution issue and others. But I want to say that finally now all the features have been made, now we have till October 3rd to hopefully get them all to be reviewed fully and ready for deployment :).

 I think most of the features would probably not be too time-consuming except maybe the first and the ruler rotation ones due to their size. I will be updating the progress of the features tomorrow.

**Comment by mohamedAdhamc** [comment]
Feature Status:

feature 1 #2717  :  Biggest thing here as of now is the history pollution which is ready for review now after the new implementation in #2750 .

feature 2 #2731  :  fully reviewed-- by ice0 and last that was said was "Looks good."

feature 3: https://github.com/synfig/synfig/pull/2755 : ice0 suggested waiting till a decision is made on whether we should
get around gtk not allowing tab to be used as an accelerator or not.

feature 4: https://github.com/synfig/synfig/pull/2634 : merged :)

feature 5: https://github.com/synfig/synfig/pull/2763 : reviewed once, and was "initial review" by rodolforg

feature 6: https://github.com/synfig/synfig/pull/2765 : This one is close to done being reviewed, but there are certain parts that need
opinions (around 2 points). as stated by rodolf
" One thing that bothers me is that the current time changes when dragging the playback boundaries. But I don't know if it is really a problem or just me XD The rest seems to work nice :) "

feature 7: https://github.com/synfig/synfig/pull/2767 :  Review by rodolforg is done and as said by him " I believe the only points left (or not) are those I'm waiting for ice0 to comment/review. "

feature 8: https://github.com/synfig/synfig/pull/2777: ready for review, not yet reviewed, and is probably the lengthiest feature out of them all.

feature 9: https://github.com/synfig/synfig/pull/2805: not yet reviewed.

feature 10: https://github.com/synfig/synfig/pull/2809, reviewed multiple times by ice0 last thing was a suggested optimization that is waiting for review.

feature 11: https://github.com/synfig/synfig/pull/2615: reviewed many times by ice0 and rodolforg I think its fullly reviewed but the last thing was a question from rodolf . "It's a Gtk theme issue. I think we should not use the separator :/ @morevnaproject @ice0".

feature 12: #2810 very small feature ice0 suggested a diffrent implementation an then reviewed it and review should be (almost done) if not completely done.

@ice0 , @rodolforg we have till October 3rd ( and also after that of course but just in terms of project time :) ) to make sure all these are fully reviewed and ready for merge I will be waiting for you reviews and will also try to go over them again myself.

**Comment by mohamedAdhamc** [comment]
updated in the same comment as of 9/14 :)
